File: client/src/main/java/io/github/hulang1024/chinesechess/scene/AbstractScene.java
package io.github.hulang1024.chinesechess.scene;

import javafx.scene.layout.Background;
import javafx.scene.layout.BackgroundFill;
import javafx.scene.layout.FlowPane;
import javafx.scene.paint.Color;

/**
 * @author Hu Lang
 */
public abstract class AbstractScene extends FlowPane {
    protected SceneContext context;
    private SceneManager sceneManager;

    public AbstractScene(SceneContext context) {
        this.context = context;
        this.sceneManager = SceneManager.of(context);

        prefWidthProperty().bind(this.context.getPrimaryStage().widthProperty());
        prefHeightProperty().bind(this.context.getPrimaryStage().heightProperty());

        setBackground(new Background(new BackgroundFill(Color.GRAY, null ,null)));
    }

    public void onSceneExit() {
        
    }

    protected void pushScene(SceneBuilder sceneBuilder) {
        onSceneExit();
        sceneManager.pushScene(sceneBuilder);
    }

    protected void popScene() {
        onSceneExit();
        sceneManager.popScene();
    }
}

File: web-client/chinese-chess/src/scene/AbstractScene.ts
import SceneContext from "./SceneContext";
import SceneManager, { SceneBuilder } from "./scene_manger";

export default abstract class AbstractScene extends eui.UILayer {
    private sceneManager: SceneManager;
    protected context: SceneContext;
    private onVisibilitychange: any;

    constructor(context: SceneContext) {
        super();
        this.context = context;
        this.sceneManager = context.sceneManager;

        document.addEventListener('visibilitychange', this.onVisibilitychange = () => {
            this.onAppVisibilityChange(document.hidden);
        });
    }

    onSceneExit() {
        document.removeEventListener('visibilitychange', this.onVisibilitychange);
    }

    onAppVisibilityChange(hidden: boolean) {}

    protected pushScene(sceneBuilder: SceneBuilder) {        
        this.sceneManager.pushScene(sceneBuilder);
    }

    protected popScene() {
        this.sceneManager.popScene();
    }
}
